warehouse/query-core/src/main/java/datawave/query/util/sortedmap/BufferedFileBackedSortedMap.java

```diff
@@ -0,0 +1,560 @@
+            }
+            map.addMap(buffer);
+        }
+        buffer.putAll(c);
+        sizeModified = true;
+        if (buffer.size() >= bufferPersistThreshold) {
+            try {
+                persist();
+            } catch (Exception ex) {
+                throw new IllegalStateException("Unable to persist or compact FileSortedMap", ex);
+            }
+        }
+    }
+
+    @Override
+    public V remove(Object o) {
+        V value = null;
+        for (SortedMap<K,V> map : map.getMaps()) {
+            FileSortedMap<K,V> filemap = (FileSortedMap<K,V>) map;
+            boolean persist = false;
+            if (filemap.isPersisted()) {
+                try {
+                    filemap.load();
+                    persist = true;
+                } catch (Exception e) {
+                    throw new IllegalStateException("Unable to remove item from underlying files", e);
+                }
+            }
+
+            V testValue = map.remove(o);
+            if (testValue != null) {
+                if (value != null) {
+                    if (rewriteStrategy == null || rewriteStrategy.rewrite((K) o, value, testValue)) {
+                        value = testValue;
+                    }
+                } else {
+                    value = testValue;
+                }
+            }
+
+            if (persist) {
+                try {
+                    filemap.persist();
+                } catch (Exception e) {
+                    throw new IllegalStateException("Unable to remove item from underlying files", e);
+                }
+            }
+        }
+        if (value != null) {
+            this.sizeModified = true;
+        }
+        return value;
+    }
+
+    @Override
+    public void clear() {
+        // This will cause the MultimapBackedSortedMap to call clear on each map in its map of maps, including the buffer
+        // It will also call clear on its map of maps, emptying the contents
+        map.clear();
+        // Null the buffer so that it will start new on the next add
+        buffer = null;
+        this.size = 0;
+        this.sizeModified = false;
+    }
+
+    @Override
+    public Comparator<K> comparator() {
+        return comparator;
+    }
+
+    @Override
+    public SortedMap<K,V> subMap(K fromKey, K toKey) {
+        return map.subMap(fromKey, toKey);
+    }
+
+    @Override
+    public SortedMap<K,V> headMap(K toKey) {
+        return map.headMap(toKey);
+    }
+
+    @Override
+    public SortedMap<K,V> tailMap(K fromKey) {
+        return map.tailMap(fromKey);
+    }
+
+    @Override
+    public K firstKey() {
+        return map.firstKey();
+    }
+
+    @Override
+    public K lastKey() {
+        return map.lastKey();
+    }
+
+    @Override
+    public Set<K> keySet() {
+        persistIfMultipleMaps();
+        return map.keySet();
+    }
+
+    @Override
+    public Collection<V> values() {
+        persistIfMultipleMaps();
+        return map.values();
+    }
```

Comment:
some getter functions are persisting before returning while others are not. Can you define what operations should require persisting and which shouldn't and why?

Reply:
Now I am thinking that the persistence should never be done automatically here.  The transform should drive that.

Reply:
Updated to not automatically persisty and explicitly persist in the transform.